### PR TITLE
Acados solver code generation improvements

### PR DESCRIPTION
**Pull request by schwieni**
The goal of this PR is to improve the exported code of acados_solver by several measures:
1) Use dynamic heap allocation instead of stack allocation, this enables solving larger problems.
2) Use of calloc (malloc and init with zeros) and only set elements which are not zero.
3) If time-steps are constant use a constant and set it directly instead of first building up an array.

This also results in smaller exported code (with larger problems it is very significant), both c-code and compiled code (binary).
Also the start-times are improved in my experiments.

### CODE REVIEW

**Comment by schwieni** [comment]
There is still a test error:
>The following tests FAILED:
>	  1 - octave_test_template_ocp_disc_dyn (Failed)
However, I believe that this PR is unrelated to octave/mex.

**Review by FreyJo — COMMENTED** [review]
Hey @schwieni 
I see that always writing all the numerical values of the NLP formulation on the stack can cause problems. It would be interesting to get a rough idea of your problems dimension.

Considering this PR:
- The failing test is introduced by a bug in this PR. The code generation (rendering the templates is available in the MEX interface, and the failing test name includes `template`. See https://docs.acados.org/matlab_octave_interface/index.html#interface-structure
- The part `If * are constant use a constant and set it directly instead of first building up an array.` was wrong in multiple ways.
- It would make a review and future debugging a lot easier if you make multiple commits instead of this one https://github.com/acados/acados/pull/726/commits/fb3c4b9681e4e07bce10753106c9c9309683ae8c
- Changing the formatting should be a separate commit. I think empty lines are compiled away anyway and don't change the size of the binary.
- this is not a `bug fix` commit to me: https://github.com/acados/acados/pull/726/commits/fb3c4b9681e4e07bce10753106c9c9309683ae8c Bug fixes should be really minimal to not introduce new bugs..
- I guess applying the changes only for the matrices would be a good first step, as their memory is growing quadratically with the dimensions and these changes might already be sufficient to make the problem work for you. Also, admittedly, the way they are initialized currently is less readable.

**Comment by schwieni** [comment]
Hey @FreyJo,
first, thank you for looking at the PR and the valuable feedback.

Regarding the problem size:
In my particular case I have not many states but many shooting nodes. Example: around 27 states (NX=18, NU=9, NY=33, NH=6) but more than 10.000 shooting nodes. That leads to a stack overflow. The main reason seems to be the int arrays under section "dimensions". However, if one has a problem with many more states and less shooting nodes it might lead to the same issue (stack overflow). 
Under Linux this can be fixed by setting the environment variable OMP_STACKSIZE to something larger (> 8MB).
Interestingly, this is only happens when I call make under the exported code directory and call the generated main_xyz executable. This generates a “Segmentation fault (core dumped)” which in my case is a stack memory overflow.
It does not happen when I run the problem under python. Maybe that is a hint to something else, too?
In my case the issue is fixed by just using heap memory for the mentioned int arrays. However, with many states larger matrices might become the problem (as you also pointed out).

Regarding the PR feedback: I think I understand most of your feedback and I agree with what you say that a bug fix should be as small as possible. Very good that you took an expert look at the things.

How should I proceed? I understand that this PR should be abandoned and I should make a new PR which separates the topics more clearly. Also, should the topics all be separate PR?

I need advice for the testing: for some reason I can not get the tests which are conducted under github when initiating a PR to run on my PC. Can you give me some hints how I can do that on my computer? I believe that would enable me to better check the issues on my PC directly.

**Comment by FreyJo** [comment]
> 10.000 shooting nodes

This is really orders of magnitude larger of what I think is typically used in optimal control.
Did you consider to use less shooting nodes and longer time steps combined with higher order integration methods?
Or is there a specific reason that you need that many shooting nodes?
One reason I have in mind for this is that you want to really ensure constraint satisfaction at that many points.

> I need advice for the testing: for some reason I can not get the tests which are conducted under github when initiating a PR to run on my PC.

Why can you not run the tests?
The ones for Matlab/ Octave are to be found here:
https://github.com/acados/acados/tree/master/examples/acados_matlab_octave/test
The files `run_*` are executed on Github actions.

> It does not happen when I run the problem under python. Maybe that is a hint to something else, too?

I don't know why that is. Maybe Python has some why to circumvent the issue?

> How should I proceed? I understand that this PR should be abandoned and I should make a new PR which separates the topics more clearly. Also, should the topics all be separate PR?

yes, I suggest to make another PR with more granular commits.
Maybe one for each of those:
- use heap for the dimension vectors in acados OCP solver template
- use heap for the matrices `Vx`, `Vu`, `W` and only assign nonzero elements in acados OCP solver template
- changes for the time varying options `time_steps`, `num_stages`, `num_steps` in acados OCP solver template

I guess these changes should be enough to make the problem work for you.

I hope that works for you.

Just FYI: I'll be on vacation next week.

File: interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c [hunk 1]
```diff
@@ -858,16 +863,39 @@ int {{ model.name }}_acados_create(nlp_solver_capsule * capsule)
...
+    {
+        ocp_nlp_in_set(nlp_config, nlp_dims, nlp_in, i, "Ts", &time_step);
+        ocp_nlp_cost_model_set(nlp_config, nlp_dims, nlp_in, i, "scaling", &time_step);
+    }
+    {%- else -%}
+    // time_steps are different
+    double time_steps[N] = {
+    {%- for j in range(end=dims.N) -%}
+        {%- if j % 10 == 0 %}
+        {% endif -%}
```
[the remark this answers is not in the document]
Comment (schwieni) [reply interfaces/acados_template/acados_template/c_templates_tera/acados_solver.in.c:890]: No, this was an idea to have a line break after 10 values. However, I will change that to also use heap memory.

**Comment by schwieni** [comment]
Hi @FreyJo,
yes, the large number of shooting nodes is indeed to not miss anything. When the memory topics are/were cleared (see also one of my past PR) and the weights are set appropriately, acados and the hpmpi solver can also handle much larger problems (N=100.000) efficiently. That is at least my observation.

I am indeed also experimenting with variable step sizes, however so far the results are not yet as good as with the full problem.

I did not do tests with much larger step sizes but since the prediction is not smooth (it has a lot of high frequency content) maybe only a multi-grid approach could help to reduce the problem size. That might be worth a try in the future.

Regarding the testing issues:
my initial idea was to run the same tests which are conducted in the Github actions on my PC.
I used the “.github/linux/run_tests.sh” but I found out that it is not so easy since it behaves differently from github. I guess all the conditions must be set up according to the Github workflow.
I would just like to test and code in an iterative manner on my PC instead of first commit, push and wait for 20 minutes for the results.

For the next PR I will try to take care of all your comments above and split the commits in the three steps you suggested. I understand that changing anything on the template space removals (the ‘-’) is not desired. That is of course fine with me and I will leave that to you.

Have a nice vacation!

**Comment by imciner2** [comment]
Just a note, variable length arrays (like what is currently used in the code) are usually discouraged in code because they aren't always very portable (MISRA actually also says not to use them in Rule 18.8). While they are technically in the C99 standard, they never made it into the C++ standard. Some projects have even gone to great lengths to remove all uses of variable length arrays (such as the Linux kernel) and instead go to the heap-memory allocation model instead.

**Comment by schwieni** [comment]
@imciner2 I am not sure what you mean by "variable length arrays" and "like what is currently used in the code". I could not find the MISRA Rule 18.8. Can you give a reference or link? Also, I could not find any variable length arrays in this PR.
Let me explain what this PR is/was about:
In the acados template file acados_solver.c.in all arrays are allocated in stack. The stack memory is usually limited to some MB. For a problem which I wanted to solve this was a limiting factor. So I changed all the array allocation from stack to heap using malloc/calloc and free.
Here, all arrays are fixed size arrays and not variable length arrays.
I am preparing a new PR which clears the issues with this PR as explained in one of the posts above.

I know that there are low-end micro-controllers available which are so limited that no proper std-c library (with malloc/free) is available. Do you have such an application at hand?

**Pull request closed without merging**